3단계 - 패키지 리팩터링

hbkuk opened the pull request:
안녕하세요. 준오 리뷰어님! 😄 

3주차 `3단계 - 패키지 리팩터링` 미션에 대한 코드 리뷰 요청드립니다.

처음으로 미션을 진행하다가 **'이번 미션 너무~~~~어렵다.'** 라고 생각했던 것 같습니다.
그래도 많은 힌트를 보고나서, 제가 할수 있는 것부터.. 하나씩 진행했던 것 같습니다.

미션을 잘 진행한 걸까요..? 하하 ㅎㅎ;;

---

이번 미션에서 제가 진행했던 리팩토링은 총 2가지 입니다.
1. `Auth` 패키지와 `Member` 패키지 분리
- 진행하면서 컴파일 에러가 발생하니깐 굉장히 두려웠던 것 같습니다.
  - 테스트 코드를 잘못 삭제하거나, 주석을 수정하지 못하거나 ... 이런 부분에서 굉장히 많은 실수가 있었습니다. 😢 
  - 실무에서 사용할 수준까지... 실력을 끌어올리고 싶은데요. 혹시 준오님께서 추천하시는 훈련방법이 있을까요? 😄 
2. `Favorite`에서 `Member`으로 탐색하는 방식을 객체 참조에서 Repository를 통한 탐색으로 변경
- 리팩토링을 잘한 걸까요? .. 감이 안옵니다. 😢 

위와 같이 최대한 양방향 참조가 발생하지 않도록 진행했습니다.
 
혹시 개선할만한 부분이 있다면, 가감없이 리뷰해 주시면 정말 감사드릴 것 같습니다.

감사합니다. 😃

hbkuk committed:
docs: 테스트 관련 application.ym에 명시된 설정을 test > resources 하위 폴더로 이동

- docs: 상기 변경사항으로 인한 기존 ActiveProfiles 설정 삭제

hbkuk committed:
fix: 이메일과 나이를 검증한다는 부분으로 유추될 수 있도록 Fixture 수정 및 테스트 코드 개선

hbkuk committed:
refactor: GithubClient 클래스의 깃허브 프로필 정보를 요청하는 메서드 리팩토링

hbkuk committed:
refactor: 기존 member 패키지에 위치했던 인증 관련 클래스를 auth 패키지로 이동

- refactor: 기존 지하철 관련 모든 클래스를 subway 하위 패키지로 이동

hbkuk committed:
refactor: 기존 Token Service Test의 모든 테스트 코드를 User Detail Service Test로 이동

hbkuk committed:
refactor: User Detail Service 클래스로 기존 Token Service의 모든 메서드 이동

hbkuk committed:
refactor: Token Service에서 기존 Member Service 관련 로직 삭제 및 협력 객체에서 제외

- refactor: Token Controller에서 User Detail Service를 협력 객체로 대체

hbkuk committed:
fix: 패키지 리팩토링 과정에서 중복 선언된 메서드 삭제 및 매개변수명 수정

hbkuk committed:
fix: 패키지 리팩토링 과정에서 수정하지 못한 Token Service 레이어 테스트 코드와 테스트 시나리오 수정

hbkuk committed:
fix: 패키지 리팩토링 과정에서 삭제된 코드 기반 깃허브 로그인 Mock 테스트 추가

hbkuk committed:
refactor: 기존 Favorite에서 Member으로 탐색하는 방식을 객체 참조에서 Repository를 통한 탐색으로 변경

- docs: 로깅 설정 수정

hbkuk committed:
docs: ReadME update

hbkuk committed:
fix: Member 도메인 객체에서 id를 비교하는 메서드 이름 변경

bactoria commented on src/main/java/nextstep/core/member/application/GithubClient.java:
```diff
@@ -23,14 +23,14 @@ public class GithubClient {
         httpHeaders.add(HttpHeaders.AUTHORIZATION, accessToken);
 
-        return restTemplate
+        return new RestTemplate()
```
앗, 지난 리뷰에서 놓친 부분이네요!

깃허브 로그인할 때마다 호출되는 메서드로 보이는데, RestTemplate 를 매번 생성하고 있네요.

생성한 RestTemplate 은 기본적으로 커넥션풀을 관리합니다. (재사용하도록 설계됨)
또한, RestTemplate 를 매번 만들면 성능저하가 발생할 수도 있어요 - [참고](https://brocess.tistory.com/265)

RestTemplate 을 빈으로 등록하면 어떨까요?

<br>

추가로 `RestTemplate` 은 Spring 5 (Boot 2) 부터 [유지모드](https://effortguy.tistory.com/214) 이기 때문에 저희팀에서는 사용하고 있지 않는데요.
Spring 5 (Boot 2) 에 추가된 `WebClient` 를 사용하고 있습니다. (WebClient는 동기/비동기 호출을 지원. 단, webflux 의존성을 추가해주어야 하는 단점이 있음.)
Spring 6.1 (Boot 3.2) 에서는 `RestClient` 가 추가되었으니 
Spring 6.1 이상의 환경이라면 RestClient를 사용하는 것도 좋을 것 같아요.

hbkuk replied:
👏 
몰랐던 부분인데, 리뷰 감사드립니다 😄 

반영하도록 하겠습니다.

bactoria replied:
@hbkuk 
WebClient와 RestClient는 이런게 있다 정도로 말씀드리려고 했는데, 제가 따로 언급은 없었군요!
그래도 미션에 잘 반영해주셨네요 💯

bactoria commented on src/main/java/nextstep/core/auth/application/UserDetailService.java:
```diff
@@ -0,0 +1,35 @@
+
+import nextstep.core.auth.application.dto.GithubProfileResponse;
+import nextstep.core.auth.application.dto.TokenResponse;
+import nextstep.core.member.application.MemberService;
```
이번 미션에서는 auth 패키지에서의 member 패키지의 의존성을 제거하는 것이 목표인데요!
여기서 member 패키지에 대한 의존이 생겼네요

의존성 역전 원칙(DIP) 을 적용하여 
auth 패키지의 UserDetailsService 를 인터페이스로 변경하고, 
이 인터페이스에 대한 구현체는 member 패키지에 두면 어떨까요?

멤버쪽에 필요한 건 아래 두가지 이므로 이를 메서드로 가지는 인터페이스를 추출해보면 좋을 것 같아요!
- memberService.findMemberByEmail(email)
- memberService.findOrCreate(email)

hbkuk replied:
앗 ...

제안해주신 내용대로 인터페이스로 추출해서 구현해보도록 하겠습니다!! 👍

bactoria commented on src/main/java/nextstep/core/subway/favorite/application/FavoriteService.java:
```diff
@@ -33,20 +32,20 @@ public FavoriteService(FavoriteRepository favoriteRepository, PathFinderService
 
     public List<FavoriteResponse> findFavorites(Member member) {
-        return convertToFavoriteResponses(favoriteRepository.findByMember(member));
+        return FavoriteConverter.convertToFavoriteResponses(favoriteRepository.findByMember(member.getId()));
```
이와 같이 코드가 길어지는 경우 가독성을 해칠 수 있는 것 같아요.
개인적으로는 아래와 같이 풀어쓰는 방식이 가독성이 좋다고 생각해요!

```suggestion
        List<Favorite> favorites = favoriteRepository.findByMember(member.getId());
        return FavoriteConverter.convertToFavoriteResponses(favorites);
```

hbkuk replied:
앞으로는 이렇게 코드를 작성하지 않을 것 같습니다. 😢 
(극단적인 라인 줄이기...는 하지 않겠다.. 메모..!!)

bactoria commented on src/main/java/nextstep/core/subway/favorite/application/FavoriteService.java:
```diff
@@ -33,20 +32,20 @@ public FavoriteService(FavoriteRepository favoriteRepository, PathFinderService
     @Transactional
     public void deleteFavorite(Long favoriteId, Member member) {
-        if (!member.isMine(findFavoriteById(favoriteId))) {
+        if (!member.isSameId(findFavoriteById(favoriteId).getMemberId())) {
```
즐겨찾기로부터 memberId 필드를 가져와서 이를 비교하는 데 사용하고 있네요. 
Favorite 에 비교하는 메서드를 추가한다면 [묻지말고 시켜라!](https://origogi.github.io/oop/%EB%AC%BB%EC%A7%80-%EB%A7%90%EA%B3%A0-%EC%8B%9C%EC%BC%9C%EB%9D%BC/) 를 적용해볼 수 있을 것 같아요.


```suggestion
        Favorite favorite = findFavoriteById(favoriteId);
        if (!favorite.isOwn(member)) {
```

hbkuk replied:
이런 피드백 너무 좋은 것 같습니다. 👍 

> 묻지말고 시켜라!

명심하겠습니다!! 😄

bactoria commented:
> 진행하면서 컴파일 에러가 발생하니깐 굉장히 두려웠던 것 같습니다.
테스트 코드를 잘못 삭제하거나, 주석을 수정하지 못하거나 ... 이런 부분에서 굉장히 많은 실수가 있었습니다. 😢
실무에서 사용할 수준까지... 실력을 끌어올리고 싶은데요. 혹시 준오님께서 추천하시는 훈련방법이 있을까요? 😄

저도 리팩토링 시 컴파일 에러를 피해가기는 정말 어려운 부분이라 공감이 되네요.

이번 UserDetailsService 를 추상화시키는 과정은 아래의 케이스와 유사한 부분이 있는데요.
- 계층 구조 아키텍처에서의 DIP를 통한 도메인을 보호하는 아키텍처로 변경 - [링크](https://blog.naver.com/PostView.nhn?blogId=yb2316&logNo=222353827299)
- 스프링 시큐리티의 UserDetailsService - [링크](https://velog.io/@yaho1024/Spring-Security-UserDetailsService-%EA%B5%AC%ED%98%84-%ED%95%B4%EB%B3%B4%EA%B8%B0)

이런 경험들이 쌓이면서 익숙해 지면서 두려움은 점점 사라지는 것 같아요.
그래서 코드를 다양한 관점으로 만들어보고, 리팩토링하는 경험이 좋다고 생각합니다!
이번 단계를 통해서도 한단계 성장하셨을거에요!

<br>

> Favorite에서 Member으로 탐색하는 방식을 객체 참조에서 Repository를 통한 탐색으로 변경
리팩토링을 잘한 걸까요? .. 감이 안옵니다. 😢

객체의 생명주기가 완전 별개인 경우에 저는 직접참조는 사용하지 않고, 간접참조로 개발을 하는데요.
Member의 생명주기는 Favorite와 완전 별개이므로 간접참조가 좋은 것 같아요!

Member와 Subway 를 별도의 패키지로 구분지어 주셨는데, 
시스템이 복잡해질 수록 Member 를 Subway와 별도의 패키지로 분리를 할 것 같고, 
더 복잡해지면 Member 를 Subway와 별도의 서버로 분리를 할 것 같고,
더 복잡해지면 Member 서버를 담당하는 팀을 별도로 구성할 것 같아요.

그러니, 현재에는 하나의 서버에 있지만 원하면 쉽게 서버를 떼어낼 수 있도록 의존성을 관리해야 하는데, 
이를 위해 패키지를 분리와 간접참조로의 변경은 아주 좋은 것 같아요. 💯 

(자바에서는 protected 접근제한자를 이용하여 다른 패키지에서의 접근을 막을 수 있으니 이를 활용해보셔도 좋을 것 같아요!)

bactoria reviewed (changes requested):
병국님 3단계 미션 잘 진행 해주셨네요 :clap: :clap:
UserDetailsService 에 대한 추상화를 해보시면 좋을 것 같은데, 
리뷰 한번 검토 해주시고 다시 리뷰 요청 주세요!

hbkuk committed:
fix: 깃허브 로그인 요청에서 사용하는 HTTP 요청을 위한 라이브러리를 RestTemplate에서 HttpClient로 변경

- fix: HttpClient로 변경으로 인한 통신 간 발생하는 Client 예외를 핸들링하기 위한 전역 예외 핸들러 메서드 수정

- docs: WebClient 의존성 추가

hbkuk committed:
refactor: 가독성이 저하되는 메서드명 수정 및 일부 메서드 인자를 변수화

hbkuk committed:
refactor: Favorite 객체에서 Member 객체를 매개변수로 받아 Member의 소유인지 확인하는 메서드를 사용하도록 수정

hbkuk committed:
refactor: auth 패키지에서 member 패키지에 대한 의존성을 갖지 않도록 구조 변경

hbkuk committed:
fix: 이메일과 비밀번호를 받아서 회원을 검증하는 메서드명 수정

- docs: 구조 변경에서 변경되지 않은 테스트 코드 및 시나리오 수정

- fix: Member Controller import 정리

bactoria reviewed (approved):
병국님 피드백 잘 반영해 주셨네요. 💯  
열심히 미션 진행하시는 모습 보면서 저도 리뷰를 즐겁게 한 것 같아요.

즐겨찾기 미션은 여기서 마무리 짓도록 하겠습니다. 
추가로 궁금한 부분 있으시면 디엠 주세요!

끝까지 화이팅입니다 🔥
